### DISCUSSION TITLE
[Proposal] support the policy to allow the end-users' clients to self-provision restricted user-accounts / buckets without the need for having a service with admin rights

### DISCUSSION THREAD

**Discussion by epoberezkin** [minio/minio#15829] [Q&A] [answered]
As per subject, following up from the conversation in slack.

Our use case is to use minio as an object storage in our messaging app (https://github.com/simplex-chat/simplex-chat#readme), to allow the users to upload the files and send them as links (currently we support sending files in chunks, in multiple messages, but it's inefficient, particularly in decentralized groups - as the file has to be sent separately to each member).

I understand that we could create a service with admin level access to manage that, but it seems worse for two reasons:
- additional moving part to develop/deploy/maintain - we would rather invest some part of this time to extend minio.
- more difficult to deploy to the end users - we allow them to use their own messaging and TURN/STUN servers, we want them to be able to self-host object storage as well, without the need to deploy an additional application service – we would simply instruct them how to deploy minio, configure the policies and the app to use their service instead of ours.

The use case seems to be wider that just our app - any mobile application developer of the app that supports the documents might benefit from being able to offer the end users hosted (and, possibly, self-hosted) storage for the documents - there may be no other reason to have application server other than to manage min.io, and if the app could self-provision, such application server would be unnecessary.

I obviously mean that minio itself would be hosted by the app developers.

For this to work, the following policy parameters need to be supported, and should only be manageable/changeable by the server owner, not by the end-users:
- limit the number of buckets per self-provisioned user account
- limit the total file storage per account
- limit the max object size per account
- set the expiration time for created objects
- possibly, allow the end-users to set expiration time, but not longer than pre-defined maximum

**Comment by klauspost** [comment, marked as answer]
Nice to see fellow AGPL projects :)

I suggest you look at the S3 API to see what would be a feasible design. It seems that you have made a design that relies on another object store. Instead I humbly suggest that you design around an S3 compatible object store. (Perhaps surprisingly) we aren't interested in having lock-in of our users by offering propriety features.

You specify your goal...

> [...] to allow the users to upload the files and send them as links [...] in decentralized groups [...]

... and you would like to place limitations on individual users (paraphrasing).

MinIO does not have any central database per design. As a consequence we cannot do a `select sum(size) ...` in any meaningful (read: realtime) way.

Therefore to do what you want you would need an external method for tracking uploads more precisely.

A few more notes for your design:

* [Presigned URLs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html) have build-in expiration. It cannot exceed 7 days as per S3 API. These seem as a reasonable option for your design.
* [ILM](https://min.io/docs/minio/linux/reference/minio-mc/mc-ilm-add.html) [(lifecycle on S3)](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html) allows to delete objects automatically after a specific time.
* Multi-bucket design is discouraged in most cases. Keep your design to single-bucket, instead use prefixes for managing separating data.

In general you should look into a server API that manages the S3 API, so you don't have to send S3 credentials to clients. I presume you already have a client/server connection. The server can then forward uploads and manage deletes directly, while keeping track of user storage use.
